File: frontend/src/api/dashboardKpis.ts
import { fetchWithAuth } from "./http.ts";

const BASE = "/api/v1/dashboard";

/**
 * KPIs personales de nómina del usuario autenticado, desde las cachés en Bono.
 *
 * Los de vacaciones son `null` cuando `disponible` es `false` (empleado sin saldo
 * sincronizado): la UI debe pintar «—», no «0 días».
 *
 * `retardos_anio` **no** sigue a `disponible`: llega aunque el saldo falte. `null` ahí
 * significa que falló su lectura; sin retardos vale 0.
 */
export type DashboardKpisResponse = {
  disponible: boolean;
  vacaciones_disponibles: number | null;
  vacaciones_tomadas_ciclo: number | null;
  vacaciones_derecho_ciclo: number | null;
  ciclo_aniversario: number | null;
  ciclo_vence: string | null;
  home_office_dias_anio: number | null;
  retardos_anio: number | null;
  anio: number;
};

/** `null` si la petición falla: el dashboard se dibuja igual, con los KPIs en «—». */
export async function fetchDashboardKpis(): Promise<DashboardKpisResponse | null> {
  try {
    const res = await fetchWithAuth(`${BASE}/mis-kpis`);
    if (!res.ok) return null;
    return (await res.json()) as DashboardKpisResponse;
  } catch {
    return null;
  }
}

File: app/schemas/dashboard_kpis.py
from datetime import date

from pydantic import BaseModel, Field


class DashboardKpisResponse(BaseModel):
    """KPIs personales de nómina (TRESS) para las tarjetas del dashboard.

    Los valores numéricos son ``None`` cuando ``disponible`` es ``False``: a diferencia
    del resto de consumidores de TRESS —que bloquean con 503 porque un saldo incorrecto
    generaría una solicitud inválida—, el dashboard degrada y la UI pinta «—». Un 0 real
    y un fallo de conexión no deben verse igual.
    """

    disponible: bool = Field(
        description="False si datos-analisis no está configurada o no respondió."
    )

    vacaciones_disponibles: float | None = Field(
        default=None, description="Suma de días de gozo pendientes de todos los aniversarios."
    )
    vacaciones_tomadas_ciclo: float | None = Field(
        default=None, description="Días ya gozados del aniversario vigente."
    )
    vacaciones_derecho_ciclo: float | None = Field(
        default=None, description="Días que corresponden en el aniversario vigente."
    )
    ciclo_aniversario: int | None = Field(
        default=None, description="Número de aniversario vigente en TRESS."
    )
    ciclo_vence: date | None = Field(
        default=None, description="Fecha de vencimiento del ciclo vigente, si la hay."
    )

    home_office_dias_anio: int | None = Field(
        default=None, description="Días de home office registrados en el año en curso."
    )
    retardos_anio: int | None = Field(
        default=None,
        description=(
            "Retardos del año en curso, desde la caché `levelup_incidencias_tress`. "
            "**No depende de `disponible`**, que describe solo el bloque de vacaciones: "
            "un empleado sin saldo sincronizado igual recibe su conteo. `None` solo si "
            "falla la lectura; sin retardos es 0."
        ),
    )
    anio: int = Field(description="Año al que corresponden los KPIs de home office y retardos.")
